[egs] Speaker verification recipe for the VoxCeleb2 corpus

entn-at opened the pull request:
This is a recipe for speaker verification using the recently published VoxCeleb2 corpus, containing over 1 million utterances for 6,114 celebrities, extracted from videos uploaded to YouTube (http://www.robots.ox.ac.uk/~vgg/data/voxceleb2/). The two included systems (v1: GMM-UBM i-vector PLDA, v2: x-vector PLDA) are almost identical to those in the SRE16 recipe. The main difference to the SRE16 recipe (aside from the condition of the data) is that the VoxCeleb2 data is freely available.

The authors have not published an official trials list for the test set. The recipe therefore generates a trials file containing comparisons between each of the 36,693 test utterances [36,693 * (36,693-1) / 2 = 673,169,778 comparisons].

Note: I'm not affiliated with the authors/institution that created the VoxCeleb2 corpus.

danpovey commented:
Thanks-- I'll wait for @david-ryan-snyder to review.

david-ryan-snyder commented:
Thanks! I'll review this soon.

I was actually just about to do something like this myself!

entn-at commented:
Thanks! I'd be happy to incorporate suggestions for improvements/extensions. I didn't tune the recipes a lot, as I only have limited compute resources. Other audio-from-video type datasets could perhaps be included, such as VoxCeleb1 and SITW (but that could also be done later in a separate PR).

david-ryan-snyder commented on egs/voxceleb2/v2/run.sh:
```diff
@@ -0,0 +1,210 @@
+
+if [ $stage -le 10 ]; then
+  # Treat each test utterance independently from the other test utterances of the same speaker
+  awk '{print $1, 1;}' <data/voxceleb2_test/utt2spk >exp/xvectors_voxceleb2_test/num_utts.ark
```
I don't think this is needed, since you can just remove the --num-utts option.

entn-at replied:
Removed.

david-ryan-snyder commented on egs/voxceleb2/v2/run.sh:
```diff
@@ -0,0 +1,210 @@
+
+  $train_cmd exp/scores/log/voxceleb2_test_scoring.log \
+    ivector-plda-scoring --normalize-length=true \
+    --num-utts=ark:exp/xvectors_voxceleb2_test/num_utts.ark \
```
I think you can remove this option.

entn-at replied:
Removed.

entn-at commented:
Thanks for the review! I'd be happy to help out with further developing this set of recipes.

Just a note: Creation of the trials file takes quite a while and results in a 39GB text file, which isn't ideal but required the least amount of changes to the usual pipeline. I thought about creating a variant of `ivector-plda-scoring-dense` that performs all unique comparisons within a set of {i,x}-vectors, but it might not be worth the trouble, given that this would be the only recipe to make use of it.

entn-at commented on egs/voxceleb2/v1/run.sh:
```diff
@@ -0,0 +1,115 @@
+if [ $stage -le 3 ]; then
+  # Train the i-vector extractor.
+  sid/train_ivector_extractor.sh --cmd "$train_cmd --mem 20G" \
+    --ivector-dim 400 --num-iters 5 --stage 1 \
```
Yes, I accidentally didn't remove it during cleanup. Removed.

entn-at commented on egs/voxceleb2/v1/run.sh:
```diff
@@ -0,0 +1,115 @@
+
+if [ $stage -le 7 ]; then
+  # Treat each test utterance independently from the other test utterances of the same speaker
+  awk '{print $1, 1;}' <data/voxceleb2_test/utt2spk >exp/ivectors_voxceleb2_test/num_utts.ark
```
Thanks, good to know. Removed.

entn-at commented on egs/voxceleb2/v1/run.sh:
```diff
@@ -0,0 +1,115 @@
+
+  $train_cmd exp/scores/log/voxceleb2_test_scoring.log \
+    ivector-plda-scoring --normalize-length=true \
+    --num-utts=ark:exp/ivectors_voxceleb2_test/num_utts.ark \
```
Removed.

entn-at commented on egs/voxceleb2/v1/local/nnet3:
```diff
@@ -0,0 +1 @@
+../../../sre16/v2/local/nnet3/
```
I copied the scripts from egs/sre16/v1/local/ (with the source file/commit specified in a comment).

entn-at commented:
@david-ryan-snyder I pushed a commit with fixes for these inconsistencies.

david-ryan-snyder commented on egs/voxceleb2/v2/run.sh:
```diff
@@ -0,0 +1,206 @@
+    "ark:ivector-subtract-global-mean scp:exp/xvectors_voxceleb2_train/xvector.scp ark:- |" \
+    ark:data/voxceleb2_train/utt2spk exp/xvectors_voxceleb2_train/transform.mat || exit 1;
+
+  # Train an out-of-domain PLDA model.
```
I think this comment is left over from earlier. This is definitely in-domain.

entn-at commented:
Thanks for reviewing the scripts so diligently, @david-ryan-snyder. I wish I had caught these things before I created the PR. The comment was copied over from the SRE16 recipe.

david-ryan-snyder commented:
No problem!

> Just a note: Creation of the trials file takes quite a while and results in a 39GB text file, which isn't ideal but required the least amount of changes to the usual pipeline.

Yes, this isn't ideal...

It'd be best if the Voxceleb2 guys create an "official" trials file we can use. I emailed them, let's wait and see if they are going to provide something--they did for the first Voxceleb (http://www.robots.ox.ac.uk/~vgg/data/voxceleb/voxceleb1_test.txt). Let's see what they say before moving forward.

entn-at commented:
Agreed, that would be best. Joon Son has been very responsive to emails in the past. I would imagine that they are going to post the accompanying paper on arXiv soon, which probably includes verification results for trials drawn from the test set.

david-ryan-snyder commented:
He says that they've been using the original VoxCeleb test set for evaluation purposes, but they might release one based on VoxCeleb2 in the future.

Do you have access to either VoxCeleb1 or SITW? I would prefer if we evaluate on the same things others are, to keep results comparable across groups.

entn-at commented:
Yes, I do. I'm going to change the name of the recipe to "voxceleb" and use the VoxCeleb1 test set for testing.

david-ryan-snyder commented:
Perfect, thanks!

entn-at commented:
@david-ryan-snyder I pushed commits with changes to use the verification trials from VoxCeleb1 for testing and removed the old code to generate VoxCeleb2 trials. I also renamed the recipe to `voxceleb`.

- GMM-UBM i-vector PLDA: EER: 5.748%
- x-vector PLDA: EER: 4.173%

I'm sure this can be improved upon.

david-ryan-snyder commented:
@entn-at, thanks!

@danpovey, can you merge this?
